### DISCUSSION TITLE
How to add styled plugin in vtsls

### DISCUSSION THREAD

**Discussion by gilmarsquinelato** [LazyVim/LazyVim#3435] [Q&A]
Hello everyone,

I'm trying to add the `@styled/typescript-styled-plugin` in my config, but it doesn't work.
When I completely overwrite the typescript config including the `globalPlugins` with it, it works as expected, giving proper autocomplete for CSS properties, etc.

But when I try to just extend the existing config, as the vue config does (https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/vue.lua#L33), it doesn't work at all.

Did anyone face this situation already?

Thanks in advance 😃

**Comment by gilmarsquinelato** [comment]
So, weirdly I got it solved by enforcing the typescript extra as required in the `lazy.lua` file, even having it already marked as installed in via the Extras UI.

then in one of my `plugins` file I added this:

```lua
  {
    "neovim/nvim-lspconfig",
    opts = function(_, opts)
      -- replace this path to the styled plugin path on your system
      local styled_plugin_path =
        vim.fn.expand("$HOME/.nvm/versions/node/v21.6.1/lib/node_modules/@styled/typescript-styled-plugin")

      LazyVim.extend(opts.servers.vtsls, "settings.vtsls.tsserver.globalPlugins", {
        {
          name = "@styled/typescript-styled-plugin",
          location = styled_plugin_path,
          enableForWorkspaceTypeScriptVersions = true,
        },
      })
    end,
  },
```

**Reply by dpetka2001** [reply]
You shouldn't have to do that. Have you tried the following
```lua
  {
    "neovim/nvim-lspconfig",
    opts = function(_, opts)
	  local styled_plugin_path =
        vim.fn.expand("$HOME/.nvm/versions/node/v21.6.1/lib/node_modules/@styled/typescript-styled-plugin")

      vim.list_extend(opts.servers.vtsls.settings.vtsls.tsserver.globalPlugins, {
        {
          name = "@styled/typescript-styled-plugin",
          location = styled_plugin_path,
          enableForWorkspaceTypeScriptVersions = true,
        },
      })
    end,
  },
```
and it didn't work?

**Reply by gilmarsquinelato** [reply]
so, I didn't try this way, I tried the same way it was done in the Vue config.
It didn't work at all, but as soon as I explicitly added the typescript extra in lazy.lua it worked, just to reiterate, I already had the typescript extra installed through the Extras UI.

I guess it was related to how the distro is dealing with the extras are loaded, if they are only added in the Extras UI, the local plugins folder is loaded first, not having the distro config loaded yet, losing our custom config. While adding them in the lazy.lua we have the extra being loaded before our custom config.

**Reply by dpetka2001** [reply]
Can you remove the entry from `lazy.lua` and type this in the command line `=LazyVim.opts("nvim-lspconfig").servers.vtsls.settings.vtsls.tsserver` and show its output? Also you should keep the `LazyVim.extend` instead of `vim.list_extend` (I suggested this assuming you had the Vue Extra enabled, but apparently that's not the case). Regarding the Extras I believe they're loaded before the user's personal config, but not 100% sure. The maintainer will know better.

**Reply by gilmarsquinelato** [reply]
So, it's weird, I dropped the config in `lazy.lua` and it worked normally 🤔 
Could it have been a problem while loading it before, and now with the enforcement it worked as expected?

I tried several times with different configs without luck, and it only worked after the `lazy.lua` config, and now even after removing the explicit dependency on typescript it's working as expected.

**Reply by gilmarsquinelato** [reply]
And as you asked, here's the output:

```lua
{
  globalPlugins = { {
      enableForWorkspaceTypeScriptVersions = true,
      location = "/Users/<user>/.nvm/versions/node/v21.6.1/lib/node_modules/@styled/typescript-styled-plugin",
      name = "@styled/typescript-styled-plugin"
    } }
}
```

**Reply by dpetka2001** [reply]
For me it's working without having to define it in `lazy.lua`. Now I don't do any Typescript, so don't know how it works in a TS file. But I tried the following
```lua
  {
    "neovim/nvim-lspconfig",
    opts = function(_, opts)
      LazyVim.extend(opts.servers.vtsls, "settings.vtsls.tsserver.globalPlugins", {
        {
          name = "@styled/typescript-styled-plugin",
          location = "styled_plugin_path",
          enableForWorkspaceTypeScriptVersions = true,
        },
      })
    end,
  },
```
and when I do `:=LazyVim.opts("nvim-lspconfig").servers.vtsls.settings.vtsls.tsserver` I can see the following output 
```lua
  globalPlugins = { {
      enableForWorkspaceTypeScriptVersions = true,
      location = "styled_plugin_path",
      name = "@styled/typescript-styled-plugin"
    } }
```
which suggests that the custom setting made it into the `vtsls` server settings.

**Reply by dpetka2001** [reply]
Don't mind the location. I made it into a string because I don't do Typescript and didn't want to set up an environment (otherwise I was getting an error).

**Reply by gilmarsquinelato** [reply]
Yes, it's solved now :)

**Reply by gilmarsquinelato** [reply]
BTW, thanks for jumping in and helping me on addressing this :)
